crates/walrus-service/src/node/storage/blob_info.rs

```diff
@@ -149,6 +161,120 @@ impl BlobInfoTable {
                 [(object_id, per_object_operation.to_bytes())],
             )?;
         }
+
+        batch.insert_batch(&latest_handled_event_index, [(&(), event_index)])?;
+        batch.write()
+    }
+
+    /// Updates the blob info for a blob based on the [`BlobEvent`] when the node is in recovery
+    /// with incomplete history.
+    ///
+    /// Only updates the info if the provided `event_index` hasn't been processed yet.
+    #[tracing::instrument(skip(self))]
+    pub fn update_blob_info_during_recovery_with_incomplete_history(
+        &self,
+        event_index: u64,
+        event: &BlobEvent,
+        first_epoch: Epoch,
+        epoch_at_start: Epoch,
+    ) -> Result<(), TypedStoreError> {
+        tracing::debug!("updating blob info during recovery with incomplete history");
+        let extension_event = match event {
+            BlobEvent::Registered(BlobRegistered { end_epoch, .. })
+            | BlobEvent::Certified(BlobCertified { end_epoch, .. })
+            | BlobEvent::Deleted(BlobDeleted { end_epoch, .. })
+                if end_epoch <= &epoch_at_start =>
+            {
+                tracing::debug!(
+                    "skip updating blob info for event with end epoch before epoch at start"
+                );
+                return Ok(());
+            }
+            BlobEvent::Registered(_)
+            | BlobEvent::Certified(BlobCertified {
+                is_extension: false,
+                ..
+            })
```

Comment:
Nit: I think it maybe worth documenting that why if a non-expiring certified event must have the corresponding registered event available.

Here is my understanding: the Certified event's epoch is the certified epoch, which maybe register event epoch +1, but the end epoch must calculated from the registered event epoch, not certified event epoch.